pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -1161,25 +1161,16 @@ private void executeKeySharedMessageListener(Message<?> message, Runnable runnab
         executorProvider.getExecutor(peekMessageKey(message)).execute(runnable);
     }
 
+    protected void callMessageListenerAndIncreasePermits(Message<T> msg) {
+        if (log.isDebugEnabled()) {
+            log.debug("[{}][{}] Calling message listener for message {}", topic, subscription, msg.getMessageId());
+        }
+        increasePermits(msg);
+        callMessageListener(msg);
+    }
+
     protected void callMessageListener(Message<T> msg) {
         try {
-            if (log.isDebugEnabled()) {
```

Comment:
What's the main fix there? only split this method into `increasePermits` and `callMessageListener`?

Reply:
it increases permits if message is discarded and listener is configured.